### ISSUE TITLE
Consider having self.submit() with no arguments passed submit the only form on the page, or last one used?

### ISSUE THREAD

**Issue by dvcolgan** [django-functest/django-functest#7]
Often there will only be a single form on a page, so it would be cool if I didn't have to specify `self.submit('input[type=submit]')` and could just do `self.submit()`.

As an added feature, I don't think you'd ever really want to fill out more than one form on a page at once, so I've played around with a feature locally that keeps track of the last form you filled out and then has `self.submit()` submit that form if there is more than one form on the page.  This would be as opposed to crashing if there are multiple possible forms to submit.

Would you be interested in me taking a stab at either or both of these?

**Comment by spookylukey** [comment]
I think this sounds like a good idea, with some conditions:

1) If there is more than one submit button for the form, it should raise an error. Otherwise you are going to have a harder time debugging when you decide to add a new button to a form.

2) It should work the same for Selenium and WebTest. In particular, with WebTest there is the option of "submitting with no particular submit button"  - http://webtest.pythonpaste.org/en/latest/forms.html#submit-a-form  . In reality, users and browsers don't do this - users press buttons, and browsers submit with default buttons (see http://stackoverflow.com/questions/925334/how-is-the-default-submit-button-on-an-html-form-determined ). We should make sure the WebTest behaviour is the same in terms of the inputs that are submitted in the HTTP request.

As always, there would need to be tests and docs for this, including the cases for multiple submit buttons and multiple forms. This will need some new views and templates creating in the tests i.e. a 'single submit button' view and a 'multiple form' view.

Looking forward to your patch!

**Comment by dvcolgan** [comment]
So to make sure we are on the same page, you are saying:

### Case 1:

If there is a single form on a page with a single submit button,
And then I use `self.submit`,
It will see only one form on the page and submit

### Case 2:

If there are two different forms on a page, each with one submit button,
I use `self.fill` and it fills out fields on the first form,
And then I use `self.submit`,
It will see that I used `self.fill` on the first form and successfully submit the first form

### Case 3:

If there is a single form on a page with more than one submit button,
And then I use `self.submit`,
It sees more than one submit button and crashes

### Case 4:

If there is more than one form on a page, and the first has multiple submit buttons,
And I use `self.fill` on the first form that has multiple submit buttons,
And then I use `self.submit`,
Even though it remembers the last form I filled, it still crashes because there are multiple submit buttons

---

Is the above correct? Everything else you said makes sense too.

**Comment by spookylukey** [comment]
Yes, that all sounds right. Obviously 'crashing' with a helpful error message for the developer would be best :-)

**Comment by spookylukey** [comment]
I'm having second thoughts about this proposal . It would introduce a kind of state fulness that might be confusing. 

Also, as of the last release, we now support doing `submit(form_css_selector) `.  This submits the form, but without choosing any submit button. This is necessary sometimes for forms without submit buttons, especially those that are normally auto submitted by Javascript. 

I'm not sure this fits well with the functionality proposed in this issue.
